### PR TITLE
Drop support of GHC < 8

### PR DESCRIPTION
**Pull request by Bodigrim**
As discussed at previous maintainers meetings. 
A utf8 journey would go smoother without worrying about ancient GHCs.

CC @chessai @emilypi

### CODE REVIEW

**Review by chessai — APPROVED** [review]

File: src/Data/Text/Internal.hs [hunk 1]
```diff
@@ -124,11 +120,11 @@ firstf _  Nothing      = Nothing
...
-#if WORD_SIZE_IN_BITS == 64
-mul a b = fromIntegral $ fromIntegral a `mul64` fromIntegral b
-#else
-mul a b = fromIntegral $ fromIntegral a `mul32` fromIntegral b
-#endif
+mul a b
+  | finiteBitSize (0 :: Word) == 64
+  = fromIntegral $ fromIntegral a `mul64` fromIntegral b
+  | otherwise
+  = fromIntegral $ fromIntegral a `mul32` fromIntegral b
```
Comment (emilypi) [inline src/Data/Text/Internal.hs:127]: This introduces some new code. Is there something wrong with keeping the machdeps import and CPP'ing these definitions?

**Review by emilypi — APPROVED** [review]
Looks like a great start. Just one non-blocking question.

File: src/Data/Text/Internal.hs (hunk 1 above)
Comment (phadej) [reply src/Data/Text/Internal.hs:127]: GHC constant folds `finiteBitSize (0 :: Word)` branches.

File: text.cabal [hunk 2]
```diff
@@ -146,35 +131,18 @@ library
...
-  else
-    build-depends: bytestring         >= 0.10.4 && < 0.12
-
   ghc-options: -Wall -fwarn-tabs -funbox-strict-fields -O2
   if flag(developer)
     ghc-options: -fno-ignore-asserts
     cpp-options: -DASSERTS
 
-  if impl(ghc >= 8.11)
-    build-depends: ghc-bignum
```
Comment (phadej) [inline text.cabal:167]: removal of this is very nice.

File: src/Data/Text/Internal.hs (hunk 1 above)
Comment (Bodigrim) [reply src/Data/Text/Internal.hs:127]: There is very little public infrastructure available to test 32 bit platforms, e. g., AFAIK no CI service has such an option. Using `finiteBitSize` allows at least to check that 32 bit code path is well typed, while with CPP it could even do not parse.

Comment (chessai) [reply src/Data/Text/Internal.hs:127]: To my knowledge, a nix one-liner is the easiest way to emulate 32-bit platforms on 64-bit ones. Simply specify the "i686-linux" system in your nixpkgs. Here is a small example illustrating this:

```
$ cat shell.nix
with (import <nixpkgs> { config = {}; overlays = []; system = "i686-linux"; });

mkShell {
  name = "32-bit-nix-emulation-example";

  GHC_INCLUDE_PATH = "${pkgs.haskell.compiler.ghc8104}/lib/ghc-8.10.4/include";

  buildInputs = [
    gcc
  ];
}

$ cat main.c
#include <stdio.h>
#include "MachDeps.h"

int main() {
  printf("%i", WORD_SIZE_IN_BITS);
}

$ nix-shell

nix-shell$ gcc -Wall -I $GHC_INCLUDE_PATH main.c -o main && ./main
32

nix-shell$ exit

$ #edit shell.nix to say `system = x86_64-linux` now

$ gcc -Wall -I $GHC_INCLUDE_PATH main.c -o main && ./main
64
```

TLDR it's sufficient to specify just "system = i686-linux" to emulate a 32-bit platform. I have used this in the past to test things like chronos (the time library) on 32-bit.

**Comment by Lysxia** [comment]
Do we want to merge that soon or should we rather wait so the next release (which has quite some stuff already) remains more widely compatible?

**Comment by Bodigrim** [comment]
Feel free to release current `master`, if you feel motivated enough. I personally would not bother. Otherwise I need this PR merged early to start working on utf8 transition.

File: src/Data/Text/Internal/Unsafe/Shift.hs [hunk 3]
```diff
@@ -1,8 +1,5 @@
 {-# LANGUAGE CPP #-}
```
Comment (Lysxia) [inline src/Data/Text/Internal/Unsafe/Shift.hs:1]: Can't we get rid of this whole file? It says it's a workaround for GHC 6... I'd be fine punting this as a later TODO though.

**Review by Lysxia — APPROVED** [review]

File: src/Data/Text/Internal/Unsafe/Shift.hs (hunk 3 above)
Comment (emilypi) [reply src/Data/Text/Internal/Unsafe/Shift.hs:1]: I'd say let's move the bigger code-change work to a follow up to unblock @Bodigrim
